pkg/schemaconv/openapi_test.go

```diff
@@ -0,0 +1,249 @@
+					// untyped. In new implementation we preserve the references
+					// and instead change the raw extension type to untyped
+					f.Type.NamedType = &untypedName
+				}
+				typ.Fields[i] = f
+			}
+
+		}
+
+		if typ.Map != nil && typ.Map.ElementType.NamedType != nil {
+			if len(typ.Map.ElementRelationship) == 0 && typ.Scalar != nil && typ.List != nil && *typ.Map.ElementType.NamedType == "__untyped_deduced_" {
+				// In old implementation arbitrary/deduced map would always also
+				// include "separable".
+				// New implementation has some code paths that dont follow that
+				// (separable is default) so always attaach separable to deduced.
+				typ.Map.ElementRelationship = schema.Separable
+			} else if *typ.Map.ElementType.NamedType == "io.k8s.apimachinery.pkg.runtime.RawExtension" {
+				// See other rawextension comment elsewhere.
+				*typ.Map.ElementType.NamedType = untypedName
+			}
+		}
+	}
+
+	for k, v := range newTypeMap {
+		revertKnownDifferences(&v.Atom)
+		newTypeMap[k] = v
+	}
+
+	newTypeMap["io.k8s.apimachinery.pkg.runtime.RawExtension"] = schema.TypeDef{
+		Name: "io.k8s.apimachinery.pkg.runtime.RawExtension",
+		Atom: schema.Atom{
+			Map: &schema.Map{
+				ElementType: schema.TypeRef{
+					NamedType: &deducedName,
+				},
+			},
+		},
+	}
+
+	legacyYaml, err := schemaToYaml(legacyTypeMap)
+	if err != nil {
+		return err
+	}
+
+	newYaml, err := schemaToYaml(newTypeMap)
+	if err != nil {
+		return err
+	}
+
+	if !reflect.DeepEqual(legacyYaml, newYaml) {
+		return errors.New(cmp.Diff(legacyYaml, newYaml))
+	}
+
+	return nil
+}
+
+func TestCRDOpenAPIConversion(t *testing.T) {
+	files, err := crdFS.ReadDir("testdata/crds")
+	require.NoError(t, err)
+	for _, entry := range files {
+		t.Run(entry.Name(), func(t *testing.T) {
+			t.Parallel()
+			contents, err := crdFS.ReadFile("testdata/crds/" + entry.Name())
+			require.NoError(t, err)
+			require.NoError(t, compareSchemaConversion(string(contents)))
+		})
+	}
+}
+
+// Using all models defined in swagger.json
+// Convert to SMD using two methods:
+//  1. Spec -> SMD
+//  2. Spec -> JSON -> gnostic -> SMD
+//
+// Compare YAML forms. We have some allowed differences...
+func TestOpenAPIImplementation(t *testing.T) {
+	require.NoError(t, compareSchemaConversion(swaggerJSON))
+}
+
+func schemaToYaml(s any) (string, error) {
+	buf := bytes.NewBufferString("")
+	if err := yaml.NewEncoder(buf).Encode(s); err != nil {
+		return "", err
+	}
+	return buf.String(), nil
+}
+
+func specToSchemaViaProtoModels(input []byte) (*schema.Schema, error) {
+	document, err := openapi_v2.ParseDocument(input)
+	if err != nil {
+		return nil, err
+	}
+
+	models, err := proto.NewOpenAPIData(document)
+	if err != nil {
+		return nil, err
+	}
+
+	newSchema, err := schemaconv.ToSchema(models)
+	if err != nil {
+		return nil, err
+	}
+
+	return newSchema, nil
+}
+
+func BenchmarkOpenAPIConversion(b *testing.B) {
+	doc := spec.Swagger{}
+	require.NoError(b, doc.UnmarshalJSON([]byte(swaggerJSON)))
```

Comment:
Should we bench the CRDs too? I feel like that'd be too many, and not sure it would add a lot.

Reply:
Maybe you don't need ALL the CRDs, just pick a few?